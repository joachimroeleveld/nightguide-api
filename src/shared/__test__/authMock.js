const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
