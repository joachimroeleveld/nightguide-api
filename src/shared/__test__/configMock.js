const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
