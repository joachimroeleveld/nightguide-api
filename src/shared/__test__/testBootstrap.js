const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
