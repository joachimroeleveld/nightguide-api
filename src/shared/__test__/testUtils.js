const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
