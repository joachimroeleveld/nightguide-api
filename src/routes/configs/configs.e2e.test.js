require('../../shared/__test__/testBootstrap');

const request = require('supertest');
const sinon = require('sinon');

const { validator } = require('../../shared/openapi');
const {
  TEST_CONFIG_1,
  TEST_CONFIG_2,
  TEST_CONFIG_3,
} = require('../../shared/__test__/fixtures');
const { resetDb } = require('../../shared/__test__/testUtils');
const configRepository = require('./configRepository');

const CONFIG_SNAPSHOT_MATCHER = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
};

const sandbox = sinon.createSandbox();

describe('configs e2e', () => {
  beforeEach(async () => {
    sandbox.restore();
    await resetDb();
  });

  describe('GET /configs', () => {
    const validateResponse = validator.validateResponse('get', '/configs');

    it('happy path', async () => {
      await configRepository.createConfig(TEST_CONFIG_1);

      const res = await request(global.app).get('/configs');

      expect(res.status).toEqual(200);
      expect(res.body.totalCount).toEqual(1);
      expect(res.body.results[0]).toMatchInlineSnapshot(
        CONFIG_SNAPSHOT_MATCHER,
        `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "Config A",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should limit the amount of results to the limit parameter', async () => {
      const config1 = await configRepository.createConfig(TEST_CONFIG_1);
      await configRepository.createConfig(TEST_CONFIG_2);

      const res = await request(global.app)
        .get('/configs')
        .query({
          limit: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.limit).toEqual(1);
      expect(res.body.results[0].id).toEqual(config1._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should skip items set in offset parameter', async () => {
      await configRepository.createConfig(TEST_CONFIG_1);
      const config2 = await configRepository.createConfig(TEST_CONFIG_2);

      const res = await request(global.app)
        .get('/configs')
        .query({
          offset: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.offset).toEqual(1);
      expect(res.body.results[0].id).toEqual(config2._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('returns a totalcount with each result', async () => {
      await configRepository.createConfig(TEST_CONFIG_1);
      await configRepository.createConfig(TEST_CONFIG_2);

      const res = await request(global.app)
        .get('/configs')
        .query({ limit: 1 });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(1);
      expect(res.body.totalCount).toBe(2);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('ids filter', async () => {
      const config1 = await configRepository.createConfig(TEST_CONFIG_1);
      const config2 = await configRepository.createConfig(TEST_CONFIG_2);
      await configRepository.createConfig(TEST_CONFIG_3);

      const ids = [config1._id.toString(), config2._id.toString()].sort();

      const res = await request(global.app)
        .get('/configs')
        .query({
          ids,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(2);
      expect(res.body.results.map(item => item.id).sort()).toEqual(ids);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('query filter', async () => {
      const config1 = await configRepository.createConfig({
        ...TEST_CONFIG_1,
        name: 'Test',
      });
      await configRepository.createConfig(TEST_CONFIG_2);

      const res = await request(global.app)
        .get('/configs')
        .query({
          query: 'Test',
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(1);
      expect(res.body.results[0].id).toEqual(config1._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /configs/:configId', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/configs/{configId}'
    );

    it('happy path', async () => {
      const config1 = await configRepository.createConfig(TEST_CONFIG_1);

      const res = await request(global.app)
        .get(`/configs/${config1._id.toString()}`)
        .send()
        .expect(200);

      expect(res.body.id).toEqual(config1._id.toString());
      expect(res.body).toMatchSnapshot(CONFIG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('simple fields', async () => {
      const config1 = await configRepository.createConfig({
        ...TEST_CONFIG_1,
        name: 'Foo',
      });

      const res = await request(global.app)
        .get(`/configs/${config1._id}`)
        .send()
        .expect(200);

      expect(res.body).toMatchSnapshot(CONFIG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('POST /configs', () => {
    const validateResponse = validator.validateResponse('post', '/configs');

    it('happy path', async () => {
      const res = await request(global.app)
        .post('/configs')
        .send(configRepository.deserialize(TEST_CONFIG_1));
      const body = res.body;

      expect(res.status).toEqual(201);
      expect(body.id).toEqual(TEST_CONFIG_1._id);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('does not allow creating docs with the same name', async () => {
      await configRepository.createConfig({
        ...TEST_CONFIG_1,
        name: 'Foo',
      });

      const res = await request(global.app)
        .post('/configs')
        .send({
          ...configRepository.deserialize(TEST_CONFIG_2),
          name: 'Foo',
        });

      expect(res.status).toEqual(500);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('PUT /configs', () => {
    const validateResponse = validator.validateResponse(
      'put',
      '/configs/{configId}'
    );

    it('happy path', async () => {
      const config1 = await configRepository.createConfig(TEST_CONFIG_1);

      const res = await request(global.app)
        .put(`/configs/${config1._id}`)
        .send({
          name: 'Bar',
        });
      const body = res.body;

      expect(res.status).toEqual(200);
      expect(body.id).toEqual(TEST_CONFIG_1._id);
      expect(body.name).toEqual('Bar');
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('DELETE /configs/:configId', () => {
    const validateResponse = validator.validateResponse(
      'delete',
      '/configs/{configId}'
    );

    it('happy path', async () => {
      const createdConfig = await configRepository.createConfig(TEST_CONFIG_1);

      const res = await request(global.app).delete(
        `/configs/${createdConfig._id}`
      );

      let config = await configRepository.getConfig(createdConfig._id);

      expect(config).toBe(null);
      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /configs/name/:configName', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/configs/name/{configName}'
    );

    it('happy path', async () => {
      const config1 = await configRepository.createConfig({
        ...TEST_CONFIG_1,
        name: 'test',
        pageSlug: 'foo/bar',
      });

      const res = await request(global.app)
        .get(`/configs/name/test`)
        .query({ pageSlug: 'foo/bar' })
        .send()
        .expect(200);

      expect(res.body.id).toEqual(config1._id.toString());
      expect(res.body).toMatchSnapshot(CONFIG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('works without pageslug', async () => {
      const config1 = await configRepository.createConfig({
        ...TEST_CONFIG_1,
        name: 'test',
        pageSlug: undefined,
      });

      const res = await request(global.app)
        .get(`/configs/name/test`)
        .send()
        .expect(200);

      expect(res.body.id).toEqual(config1._id.toString());
      expect(res.body).toMatchSnapshot(CONFIG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });
  });
});
