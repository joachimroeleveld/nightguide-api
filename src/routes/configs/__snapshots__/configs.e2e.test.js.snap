// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configs e2e GET /configs/:configId happy path 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "Config A",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`;

exports[`configs e2e GET /configs/:configId simple fields 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "Foo",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`;

exports[`configs e2e GET /configs/name/:configName happy path 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "test",
  "pageSlug": "foo/bar",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`;

exports[`configs e2e GET /configs/name/:configName no pageslug 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "test",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`;

exports[`configs e2e GET /configs/name/:configName works without pageslug 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": "test",
  "payload": Object {
    "setting": 1,
  },
  "updatedAt": Any<String>,
}
`;
